search/es6/engine.go

```diff
@@ -0,0 +1,154 @@
+package es6
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/elastic/go-elasticsearch/v6"
+	"github.com/elastic/go-elasticsearch/v6/esapi"
+)
+
+type M map[string]any
+
+type responseBody struct {
+	Hits struct {
+		Total int `json:"total"`
+		Hits  []struct {
+			ID     string          `json:"_id"`
+			Source json.RawMessage `json:"_source"`
+		} `json:"hits"`
+	} `json:"hits"`
+}
+
+type Config struct {
+	Conn []string
+}
+
+type engine struct {
+	client *elasticsearch.Client
+}
+
+func NewEngine(c Config) (*engine, error) {
+	client, err := elasticsearch.NewClient(elasticsearch.Config{
+		Addresses: c.Conn,
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("elastic search: can't connect with server: %w", err)
+	}
+
+	return &engine{
+		client: client,
+	}, nil
+}
+
+func (s *engine) search(index string, requestBody M, responseBody any) error {
+	sorts := make([]string, 0)
+
+	if v, exists := requestBody["sort"]; exists {
+		sorts = append(sorts, v.([]string)...)
+		delete(requestBody, "sort")
+	}
+
+	opts := []func(*esapi.SearchRequest){
+		s.client.Search.WithContext(context.Background()),
+		s.client.Search.WithIndex(index),
+		s.client.Search.WithTrackTotalHits(true),
+		s.client.Search.WithSort(sorts...),
+	}
+
+	var buf bytes.Buffer
+	if err := json.NewEncoder(&buf).Encode(requestBody); err != nil {
+		return err
+	}
+	opts = append(opts, s.client.Search.WithBody(&buf))
+
+	res, err := s.client.Search(opts...)
+
+	if err != nil {
+		return err
+	}
+
+	defer res.Body.Close()
+
+	if res.IsError() {
+		buf := &bytes.Buffer{}
+		if _, err := io.Copy(buf, res.Body); err != nil {
+			return err
+		}
+		return errors.New("elastic search: error response: " + buf.String())
+	}
+
+	if err := json.NewDecoder(res.Body).Decode(responseBody); err != nil {
+		return fmt.Errorf("elastic search: error parsing the response body: %w", err)
+	}
+
+	return nil
+}
+
+func (s *engine) SuggestProjects(index, q string) (map[string]json.RawMessage, error) {
```

Comment:
just Search?

Reply:
This is a specific query geared towards autocomplete / suggesting projects based on a fuzzy text search.